yixxas

high

# Epoch is considered both started and not started when `block.timestamp == epochConfig[_id].epochBegin`

## Summary
We can see in the 2 modifiers below that if `block.timestamp == epochConfig[_id].epochBegin`, it will pass both checks. This can lead to various problems in the protocol such as loss of assets of users.

```solidity
modifier epochHasNotStarted(uint256 _id) {
	if (block.timestamp > epochConfig[_id].epochBegin)
		revert EpochAlreadyStarted();
	_;
}

/** @notice You can only call functions that use this modifier after the epoch has started
 */
modifier epochHasStarted(uint256 _id) {
	if (block.timestamp < epochConfig[_id].epochBegin)
		revert EpochNotStarted();
	_;
}
```

We explain why allowing a timestamp to pass both modifiers is a problem below.

## Vulnerability Detail

`resolveEpoch()` is the only function that uses the modifier `epochHasStarted`

`resolveEpoch()` can only be called once, and only when either epoch has ended due to a depeg, epoch ended without a depeg, or when depeg ends due to one side of the vault having no funds.

Essentially, this means that `resolveEpoch()` should only be called once epoch has started, and the 3 events above can only happen after epoch has started.

The issue here is that a depeg event can happen immediately after epoch has started. This will result in 2 situations that lead to loss of user funds if deposits are done when `block.timestamp == epochConfig[_id].epochBegin`.

1. Users can only deposit funds when an epoch has not started, either via a direct deposit or via a queue both guarded by the `epochHasNotStarted` modifier. If a user deposits funds to the risk vault when `block.timestamp == epochConfig[_id].epochBegin`, their funds can be immediately lost as `triggerDepeg` can be called in the same block. This should not be possible as an epoch that has not started should not be considered started as well.

2. If user deposit funds in the same block after `triggerDepeg` happens, then these funds will all be stuck in the contract as it does not even transfer to the risk side of the vault since `resolveEpoch` can only be called once.

## Impact
An important invariant is broken here which can result in loss of funds to users.

## Code Snippet
https://github.com/sherlock-audit/2023-03-Y2K/blob/main/Earthquake/src/v2/VaultV2.sol#L432-L444

## Tool used

Manual Review

## Recommendation
Consider changing `epochHasNotStarted` modifier as follows

```solidity
modifier epochHasNotStarted(uint256 _id) {
-	if (block.timestamp > epochconfig[_id].epochbegin)
+	if (block.timestamp >= epochconfig[_id].epochbegin)
		revert EpochAlreadyStarted();
	_;
}
```

